File: src/main/java/io/icednut/algorithm/exercise/graph/leet841/Solution.java
package io.icednut.algorithm.exercise.graph.leet841;

import java.util.HashSet;
import java.util.List;

class Solution {

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        var seen = new HashSet<Integer>();

        dfs(0, rooms, seen);

        return seen.size() == rooms.size();
    }

    private void dfs(Integer index, List<List<Integer>> rooms, HashSet<Integer> seen) {
        List<Integer> neighbors = rooms.get(index);

        if (!seen.contains(index)) {
            seen.add(index);

            neighbors.forEach(neighbor -> dfs(neighbor, rooms, seen));
        }
    }
}

File: src/main/kotlin/io/icednut/algorithm/exercise/graph/leet841/Solution.kt
package io.icednut.algorithm.exercise.graph.leet841

import java.util.*

class Solution {

    fun canVisitAllRooms(rooms: List<List<Int>>): Boolean {
        val stack = Stack<Int>()
        val seen = HashSet<Int>()
        val result = ArrayList<Int>()

        if (rooms.isNotEmpty()) {
            result.add(0)
            result.addAll(rooms[0])
            seen.add(0)
        }

        for (currentRoomKey in rooms.indices) {
            val roomKeys = rooms[currentRoomKey]

            if (roomKeys.isEmpty()) {
                seen.add(currentRoomKey)
            } else {
                for (roomKey in roomKeys) {
                    stack.push(roomKey)
                    seen.add(roomKey)
                }
            }

            while(true) {
                val targetRoom = if (stack.isEmpty()) {
                    null
                } else {
                    stack.pop()
                }

                if (targetRoom != null) {
                    val traverseTargetRooms = rooms[targetRoom]

                    for (roomKey in traverseTargetRooms) {
                        if (seen.contains(roomKey)) {
                            continue
                        }
                        stack.push(roomKey)
                        seen.add(roomKey)
                        result.add(roomKey)
                    }
                } else {
                    break
                }
            }

        }

        return result.containsAll(seen)
    }
}